app/src/main/java/ch/hikemate/app/model/route/HikesViewModel.kt

```diff
@@ -0,0 +1,1166 @@
+        val waypoints: List<LatLong> =
+            mutableWaypoints
+                ?: run {
+                  onFailure()
+                  return@withContext
+                }
+
+        // Launch a request for the elevation data of the hike
+        val elevation: List<Double>
+        try {
+          elevation = getElevationRepoWrapper(waypoints, hikeId)
+        } catch (e: Exception) {
+          Log.e(LOG_TAG, "Error encountered while retrieving elevation", e)
+          onFailure()
+          return@withContext
+        }
+
+        // Update the hike's elevation data
+        success = false
+        _hikesMutex.withLock {
+          val hikeFlow = _hikeFlowsMap[hikeId] ?: return@withLock
+          val hike = hikeFlow.value
+
+          val updatedHike = hike.copy(elevation = DeferredData.Obtained(elevation))
+          hikeFlow.value = updatedHike
+
+          // Update the selected hike if necessary
+          updateSelectedHike()
+
+          success = true
+        }
+
+        if (success) {
+          onSuccess()
+        } else {
+          onFailure()
+        }
+      }
+
+  /**
+   * The [ElevationService] interface has been developed without coroutines in mind, hence we need a
+   * wrapper to "convert" the [ElevationService.getElevation] function that uses callback to a
+   * suspend function.
+   */
+  private suspend fun getElevationRepoWrapper(
+      coordinates: List<LatLong>,
+      hikeId: String
+  ): List<Double> = suspendCoroutine { continuation ->
+    elevationRepo.getElevation(
+        coordinates = coordinates,
+        hikeID = hikeId,
+        onSuccess = { elevation -> continuation.resume(elevation) },
+        onFailure = { exception -> continuation.resumeWithException(exception) })
+  }
+
+  private suspend fun computeDetailsForAsync(
+      hikeId: String,
+      onSuccess: () -> Unit,
+      onFailure: () -> Unit
+  ) =
+      withContext(dispatcher) {
+        // Retrieve the current value of the hike and note what computations are required
+        val (performRequest, waypoints, elevation) =
+            checkDetailsCanBeRetrievedForAsync(hikeId, onSuccess, onFailure)
+
+        if (!performRequest) {
+          return@withContext
+        }
+
+        // Perform the actual computations
+        val distance: Double
+        val elevationGain: Double
+        val estimatedTime: Double
+        val difficulty: HikeDifficulty
+        try {
+          distance = RouteUtils.computeTotalDistance(waypoints)
+          elevationGain = RouteUtils.calculateElevationGain(elevation)
+          estimatedTime = RouteUtils.estimateTime(distance, elevationGain)
+          difficulty = RouteUtils.determineDifficulty(distance, elevationGain)
+        } catch (e: Exception) {
+          Log.e(LOG_TAG, "Error while computing hike details", e)
+
+          // Set the detail properties back to NotRequested
+          _hikesMutex.withLock {
+            val hikeFlow = _hikeFlowsMap[hikeId] ?: return@withLock
```

Comment:
I could extract it, but it seems a bit artificial to me :

- The hike flow might have been replaced during the computation while the lock was released, assuming the hike flow does not change is fair but not certain
- There is no particular gain in performance, since getting the hike flow from a map is close to $O(1)$ in expected time complexity
- It would downgrade readability in my opinion
- The util function needs to retrieve the hike flow too, and it's not ideal to pass it as a parameter

Reply:
Mmmmh. You've got great arguments. I agree on this